RFC: Timeline-based Alumni Page

yechs opened the pull request:
Resolves #71
Resolves #69

SamChou19815 committed:
RFC: A new team page with members and alumnus

Implement a members + alumni section with [Docusaurus v1](https://docusaurus.io/) project showcase as an example, but in another page.

The design is tentative.

Do not merge until we figure out these problems:

1. what to do when an alumni don't have a website
2. find enough member to replace my `sam zhou` placeholder

yechs committed:
Alumni Timeline Page: Skeleton

Building data structure & logic for the timeline component
Adding sample data with different graduating class

yechs committed:
Alumni Timeline Page: Styles

Specify the styles of components w/ CSS-Module

yechs commented:
Ah... merge conflict... should have known

yechs commented on src/components/AlumniTimeline.tsx:
```diff
@@ -0,0 +1,57 @@
+  )
+}
+
+const AlumniTimeline = ({alumni}) => {
```
This file is effectively JavaScript under the name TypeScript (since I'm not so familiar with it). Will specify the types later.

Also, the logic in this function is incredibly ugly... @SamChou19815 do you have any suggestions or ideas to make it more elegant?

SamChou19815 commented on src/components/AlumniTimeline.tsx:
```diff
@@ -0,0 +1,57 @@
+    }
+    alumniOfClass.push(alumnus);
+    lastClassOf = alumnus.classOf;
+  });
```
The alumni list partitioning step should happen outside of the component. Then the component can use `.map` to render without an imperative `forEach`

yechs replied:
Thought so, but the `alumni` array is received through `props` and is restricted in the functional scope.

Or perhaps you mean declaring a function to do the partitioning job?

SamChou19815 replied:
Instead of receiving the flattened `alumni` array, you could choose to receive a partitioned one.
For example
```tsx
// members.tsx
type AlumnusFromYear = { readonly year: number; readonly members: readonly Member[] };
const partition = (alumnus: readonly Member[]): readonly AlumnusFromYear[] => { ... };

<AlumniTimeline partitionedAlumniList={partition(alumnus)} />

// AlumniTimeline.tsx
return (
  <div>
    {partitionedAlumniList.map(({ year, members }) => (
      <div>
        {year}
        {members.map(...)}
    ))}
  </div>
)
```

yechs replied:
Yeah! Got it

SamChou19815 commented:
The design looks good to me

yechs committed:
Alumni Timeline Page: Data structure change

Changes the structure of original `alumni` variable, which greatly
simplifies the code
Thanks to @SamChou19815 !!!

yechs committed:
Alumni Timeline Page: minor style changes

Changes the styles of "Our Alumni" header
Alters the meta description tag
Adds notice to change info
**prefixes vendor CSS**, since I cannot seem to get postcss working in
docusaurus

yechs committed:
Alumni Timeline Page: Removing extra placeholders

yechs commented:
Fixed as proposed
Removed the extra placeholders to be production ready

One last problem, I couldn't seem to get postcss working, so I had to manually add the vendor prefixes. Maybe we can find a solution later with another PR

yechs committed:
Merge branch 'master' into alumni-page/sy-design

SamChou19815 reviewed (approved):
Looks good. Do you have plan for a separate members page?

yechs commented:
> Looks good. Do you have plan for a separate members page?

Actually I think current members might look better on the home page
